File: app/protocol.py
"""WS wire protocol: Command (frontend->backend, Ack'd by `id`) and Event (backend->frontend) are discriminated unions on a literal field, so malformed/unknown messages fail validation instead of being silently misinterpreted."""
from __future__ import annotations

from typing import Annotated, Literal, Optional, Union
from uuid import uuid4

from pydantic import BaseModel, Field

#: Mirrors app.drivers.base.DeviceCategory; "nozzle" is deliberately absent -- it's just two motor devices grouped at authoring time, not a runtime category.
DeviceType = Literal["valve", "motor", "light"]
Subsystem = Literal["valve", "motor", "light", "daemon"]
Severity = Literal["info", "warning", "critical"]
ZoneState = Literal["stopped", "connecting", "ready", "playing", "paused", "error"]


# ---------------------------------------------------------------------------
# Commands: frontend -> backend
# ---------------------------------------------------------------------------

class _CommandBase(BaseModel):
    id: str = Field(default_factory=lambda: uuid4().hex)


class ConnectZone(_CommandBase):
    """Reconnects already-configured instances (they connect at add-time via ADD_DRIVER_INSTANCE, so this is mainly a drop-retry); empty `device_types` means all."""
    command: Literal["CONNECT_ZONE"] = "CONNECT_ZONE"
    zone_id: int
    device_types: list[DeviceType] = []


class DisconnectZone(_CommandBase):
    command: Literal["DISCONNECT_ZONE"] = "DISCONNECT_ZONE"
    zone_id: int


class RenameZone(_CommandBase):
    """Purely a display label -- never persisted elsewhere, never read by playback; `name=None`/empty reverts to the bare zone_id."""
    command: Literal["RENAME_ZONE"] = "RENAME_ZONE"
    zone_id: int
    name: Optional[str] = None


class DeleteZone(_CommandBase):
    """Disconnects everything and removes the zone_id itself -- the only way to actually delete a zone (they're otherwise never destroyed once created; see get_zone in main.py)."""
    command: Literal["DELETE_ZONE"] = "DELETE_ZONE"
    zone_id: int


class AddDriverInstance(_CommandBase):
    """`driver_type` must be one of the values from GET /drivers; `config` is validated against that driver's own schema."""
    command: Literal["ADD_DRIVER_INSTANCE"] = "ADD_DRIVER_INSTANCE"
    zone_id: int
    instance_id: str
    driver_type: str
    config: dict


class RemoveDriverInstance(_CommandBase):
    command: Literal["REMOVE_DRIVER_INSTANCE"] = "REMOVE_DRIVER_INSTANCE"
    zone_id: int
    instance_id: str


class AddDevice(_CommandBase):
    """`nozzle_group`/`nozzle_inverter` are purely an authoring-time label for the HMI's Nozzles tab (one fountain nozzle = two VFDs moving together, per the reference PyQt6 app's Nozzle model) -- the daemon still sees two independent motor devices and playback is unaffected."""
    command: Literal["ADD_DEVICE"] = "ADD_DEVICE"
    zone_id: int
    device_id: str
    instance_id: str
    channel: str
    nozzle_group: Optional[str] = None
    nozzle_inverter: Optional[int] = None


class RemoveDevice(_CommandBase):
    command: Literal["REMOVE_DEVICE"] = "REMOVE_DEVICE"
    zone_id: int
    device_id: str


class PlayScenario(_CommandBase):
    command: Literal["PLAY_SCENARIO"] = "PLAY_SCENARIO"
    zone_id: int
    scenario_id: str


class StopZone(_CommandBase):
    command: Literal["STOP_ZONE"] = "STOP_ZONE"
    zone_id: int


class PauseZone(_CommandBase):
    command: Literal["PAUSE_ZONE"] = "PAUSE_ZONE"
    zone_id: int


class SeekZone(_CommandBase):
    command: Literal["SEEK_ZONE"] = "SEEK_ZONE"
    zone_id: int
    position: float = Field(ge=0)


class SetLoop(_CommandBase):
    command: Literal["SET_LOOP"] = "SET_LOOP"
    zone_id: int
    enabled: bool


class SetGlobalBrightness(_CommandBase):
    command: Literal["SET_GLOBAL_BRIGHTNESS"] = "SET_GLOBAL_BRIGHTNESS"
    zone_id: int
    value: int = Field(ge=0, le=100)


class SetGlobalSpeed(_CommandBase):
    command: Literal["SET_GLOBAL_SPEED"] = "SET_GLOBAL_SPEED"
    zone_id: int
    value: int = Field(ge=0, le=100)


class EmergencyStop(_CommandBase):
    command: Literal["EMERGENCY_STOP"] = "EMERGENCY_STOP"
    zone_id: Optional[int] = None  # None = all zones


class ReconnectInstance(_CommandBase):
    """Forces an immediate reconnect instead of waiting for the driver's own watchdog retry timer; outcome surfaces via ConnectionStateEvent/HardwareErrorEvent, not this command's Ack -- hardware being unreachable isn't itself a bad command."""
    command: Literal["RECONNECT_INSTANCE"] = "RECONNECT_INSTANCE"
    zone_id: int
    instance_id: str


class SetDeviceState(_CommandBase):
    """Applies `parameters` to one device immediately, bypassing the scenario player, using the same shape a scenario event for that device's category would use (see ScenarioEventDto)."""
    command: Literal["SET_DEVICE_STATE"] = "SET_DEVICE_STATE"
    zone_id: int
    device_id: str
    parameters: dict


class ResetMotorFault(_CommandBase):
    """Sends CMD_FAULT_RESET to clear a tripped VFD fault -- the reference DegDrive DGI900 stays latched until this is sent, otherwise the only recourse is power-cycling it by hand."""
    command: Literal["RESET_MOTOR_FAULT"] = "RESET_MOTOR_FAULT"
    zone_id: int
    device_id: str


Command = Annotated[
    Union[
        ConnectZone,
        DisconnectZone,
        RenameZone,
        DeleteZone,
        AddDriverInstance,
        RemoveDriverInstance,
        AddDevice,
        RemoveDevice,
        PlayScenario,
        StopZone,
        PauseZone,
        SeekZone,
        SetLoop,
        SetGlobalBrightness,
        SetGlobalSpeed,
        EmergencyStop,
        ReconnectInstance,
        SetDeviceState,
        ResetMotorFault,
    ],
    Field(discriminator="command"),
]
"""Every inbound WS frame must parse as this; `id` is optional in (server fills a default) and echoed back on the Ack so the client can correlate request/response."""


# ---------------------------------------------------------------------------
# Acks: backend -> frontend, one per received command, correlated by id
# ---------------------------------------------------------------------------

class Ack(BaseModel):
    type: Literal["ack"] = "ack"
    id: str
    ok: bool
    error: Optional[str] = None


# ---------------------------------------------------------------------------
# Events: backend -> frontend, unsolicited, streamed continuously
# ---------------------------------------------------------------------------

class ZoneStatusEvent(BaseModel):
    type: Literal["zone_status"] = "zone_status"
    zone_id: int
    state: ZoneState
    scenario_id: Optional[str] = None
    position: float = 0.0
    duration: float = 0.0
    # Reported so a reconnecting HMI or second window can learn the loop state instead of assuming off.
    is_looping: bool = False


class DeviceStateEvent(BaseModel):
    """`device_id` is an internal id fabricated by the low-level controller (NOT the operator-facing ADD_DEVICE id) -- match on `instance_id`+`channel` against GET /zones instead."""
    type: Literal["device_event"] = "device_event"
    zone_id: int
    device_id: str
    instance_id: str
    channel: str
    device_type: DeviceType
    state: dict  # e.g. {"on": true} / {"frequency": 32.5, "active": true} / {"r":255,"g":0,"b":0}


class ConnectionStateEvent(BaseModel):
    type: Literal["connection_state"] = "connection_state"
    zone_id: int
    subsystem: Subsystem
    connected: bool
    detail: Optional[str] = None


class HardwareErrorEvent(BaseModel):
    """Modbus drop, Art-Net send failure, etc. Always survivable — never a crash."""
    type: Literal["hardware_error"] = "hardware_error"
    zone_id: Optional[int]
    subsystem: Subsystem
    message: str
    severity: Severity = "warning"


class HeartbeatEvent(BaseModel):
    """Lets the frontend distinguish 'daemon alive, nothing happening' from a dead socket."""
    type: Literal["heartbeat"] = "heartbeat"
    ts: float


Event = Annotated[
    Union[
        ZoneStatusEvent,
        DeviceStateEvent,
        ConnectionStateEvent,
        HardwareErrorEvent,
        HeartbeatEvent,
    ],
    Field(discriminator="type"),
]

File: src/renderer/src/lib/protocol.ts
/** Hand-written mirror of fountain-daemon/app/protocol.py -- keep in sync manually unless the protocol outgrows hand-syncing, then generate from the Pydantic schema instead. */

// Matches app.drivers.base.DeviceCategory; "nozzle" is just two motor devices grouped at authoring time, not a real category.
export type DeviceType = "valve" | "motor" | "light";
export type Subsystem = DeviceType | "daemon";
export type Severity = "info" | "warning" | "critical";
export type ZoneState = "stopped" | "connecting" | "ready" | "playing" | "paused" | "error";

// -- Commands: renderer -> daemon -------------------------------------------

interface CommandBase {
  id?: string; // optional on the way out; the daemon fills a default if omitted
}

export interface ConnectZoneCommand extends CommandBase {
  command: "CONNECT_ZONE";
  zone_id: number;
  device_types?: DeviceType[];
}

export interface DisconnectZoneCommand extends CommandBase {
  command: "DISCONNECT_ZONE";
  zone_id: number;
}

export interface RenameZoneCommand extends CommandBase {
  command: "RENAME_ZONE";
  zone_id: number;
  name: string | null;
}

export interface DeleteZoneCommand extends CommandBase {
  command: "DELETE_ZONE";
  zone_id: number;
}

export interface AddDriverInstanceCommand extends CommandBase {
  command: "ADD_DRIVER_INSTANCE";
  zone_id: number;
  instance_id: string;
  driver_type: string;
  config: Record<string, unknown>;
}

export interface RemoveDriverInstanceCommand extends CommandBase {
  command: "REMOVE_DRIVER_INSTANCE";
  zone_id: number;
  instance_id: string;
}

export interface AddDeviceCommand extends CommandBase {
  command: "ADD_DEVICE";
  zone_id: number;
  device_id: string;
  instance_id: string;
  channel: string;
  nozzle_group?: string | null;
  nozzle_inverter?: 1 | 2 | null;
}

export interface RemoveDeviceCommand extends CommandBase {
  command: "REMOVE_DEVICE";
  zone_id: number;
  device_id: string;
}

export interface PlayScenarioCommand extends CommandBase {
  command: "PLAY_SCENARIO";
  zone_id: number;
  scenario_id: string;
}

export interface StopZoneCommand extends CommandBase {
  command: "STOP_ZONE";
  zone_id: number;
}

export interface PauseZoneCommand extends CommandBase {
  command: "PAUSE_ZONE";
  zone_id: number;
}

export interface SeekZoneCommand extends CommandBase {
  command: "SEEK_ZONE";
  zone_id: number;
  position: number;
}

export interface SetLoopCommand extends CommandBase {
  command: "SET_LOOP";
  zone_id: number;
  enabled: boolean;
}

export interface SetGlobalBrightnessCommand extends CommandBase {
  command: "SET_GLOBAL_BRIGHTNESS";
  zone_id: number;
  value: number;
}

export interface SetGlobalSpeedCommand extends CommandBase {
  command: "SET_GLOBAL_SPEED";
  zone_id: number;
  value: number;
}

export interface EmergencyStopCommand extends CommandBase {
  command: "EMERGENCY_STOP";
  zone_id?: number | null;
}

/** Clears a tripped VFD fault so the drive accepts run commands again; only meaningful for a motor device_id, else the daemon Acks(ok=false). */
export interface ResetMotorFaultCommand extends CommandBase {
  command: "RESET_MOTOR_FAULT";
  zone_id: number;
  device_id: string;
}

/** Forces an immediate reconnect instead of waiting for the watchdog; outcome arrives via ConnectionStateEvent/HardwareErrorEvent, not this command's ok/error. */
export interface ReconnectInstanceCommand extends CommandBase {
  command: "RECONNECT_INSTANCE";
  zone_id: number;
  instance_id: string;
}

/** Applies parameters to one device immediately, bypassing the scenario player (used by the Devices tab); same parameter shape as a scenario event for that category. */
export interface SetDeviceStateCommand extends CommandBase {
  command: "SET_DEVICE_STATE";
  zone_id: number;
  device_id: string;
  parameters: Record<string, unknown>;
}

export type Command =
  | ConnectZoneCommand
  | DisconnectZoneCommand
  | RenameZoneCommand
  | DeleteZoneCommand
  | AddDriverInstanceCommand
  | RemoveDriverInstanceCommand
  | AddDeviceCommand
  | RemoveDeviceCommand
  | PlayScenarioCommand
  | StopZoneCommand
  | PauseZoneCommand
  | SeekZoneCommand
  | SetLoopCommand
  | SetGlobalBrightnessCommand
  | SetGlobalSpeedCommand
  | EmergencyStopCommand
  | ReconnectInstanceCommand
  | SetDeviceStateCommand
  | ResetMotorFaultCommand;

// -- Ack: daemon -> renderer, one per command, correlated by id -------------

export interface Ack {
  type: "ack";
  id: string;
  ok: boolean;
  error: string | null;
}

// -- Events: daemon -> renderer, streamed continuously -----------------------

export interface ZoneStatusEvent {
  type: "zone_status";
  zone_id: number;
  state: ZoneState;
  scenario_id: string | null;
  position: number;
  duration: number;
  is_looping: boolean;
}

export interface DeviceStateEvent {
  type: "device_event";
  zone_id: number;
  /** NOT the operator-facing device_id from GET /zones -- an internal id the controller fabricates; correlate via `instance_id` + `channel` instead. */
  device_id: string;
  instance_id: string;
  channel: string;
  device_type: DeviceType;
  state: Record<string, unknown>;
}

export interface ConnectionStateEvent {
  type: "connection_state";
  zone_id: number;
  subsystem: Subsystem;
  connected: boolean;
  detail: string | null;
}

export interface HardwareErrorEvent {
  type: "hardware_error";
  zone_id: number | null;
  subsystem: Subsystem;
  message: string;
  severity: Severity;
}

export interface HeartbeatEvent {
  type: "heartbeat";
  ts: number;
}

export type DaemonEvent =
  | ZoneStatusEvent
  | DeviceStateEvent
  | ConnectionStateEvent
  | HardwareErrorEvent
  | HeartbeatEvent;

/** Anything the daemon can send on the wire: an Ack, or one of the Event types. */
export type IncomingMessage = Ack | DaemonEvent;

export function isAck(msg: IncomingMessage): msg is Ack {
  return msg.type === "ack";
}

// -- REST: GET /drivers -------------------------------------------------------

/** One entry per registered driver; render the add-device form from `config_schema` rather than hardcoding fields per driver_type. */
export interface DriverDescriptorDto {
  driver_type: string;
  category: DeviceType;
  display_name: string;
  config_schema: Record<string, unknown>;
}

// -- REST: GET /zones ----------------------------------------------------------

export interface DriverInstanceDto {
  instance_id: string;
  driver_type: string;
  category: DeviceType;
  connected: boolean;
  config: Record<string, unknown>;
}

export interface DeviceDto {
  device_id: string;
  instance_id: string;
  channel: string;
  category: DeviceType;
  // Only meaningful for a motor device that's one of a nozzle's two inverters; null/undefined otherwise.
  nozzle_group?: string | null;
  nozzle_inverter?: 1 | 2 | null;
}

export interface ZoneConfigDto {
  zone_id: number;
  /** Display label only -- unset zones fall back to "Zone {id}" everywhere. */
  name?: string | null;
  driver_instances: DriverInstanceDto[];
  devices: DeviceDto[];
  /** Runtime-only multipliers (0-100), never persisted -- lets the HMI sliders learn the real value on load/reconnect instead of assuming 100. */
  global_brightness: number;
  global_speed: number;
}

// -- REST: GET /scenarios -------------------------------------------------------

export interface ScenarioDto {
  scenario_id: string;
  name: string;
  duration: number;
}

// -- REST: /schedule -----------------------------------------------------------

/** `time` is "HH:MM" 24h local; `days` is 0=Monday..6=Sunday, empty = every day. */
export interface ScheduleEntryDto {
  id: string;
  zone_id: number;
  scenario_id: string;
  time: string;
  days: number[];
  enabled: boolean;
  last_fired_date: string | null;
}

export interface ScheduleEntryInput {
  zone_id: number;
  scenario_id: string;
  time: string;
  days: number[];
  enabled: boolean;
}

// -- REST: GET /audit -----------------------------------------------------------

/** Answers "what happened", not "who" -- there's no operator-identity system on this shared panel. */
export interface AuditLogEntryDto {
  ts: string;
  command: string;
  zone_id: number | null;
  ok: boolean;
  error: string | null;
}
